src/org/zaproxy/zap/extension/frontendscanner/ExtensionFrontEndScanner.java

```diff
@@ -0,0 +1,380 @@
+        registerUserScripts(passiveScriptType);
+    }
+
+    @Override
+    public boolean canUnload() {
+        // The extension can be dynamically unloaded, all resources used/added can be freed/removed from core.
+        return true;
+    }
+
+    @Override
+    public void unload() {
+        super.unload();
+
+        this.extensionScript.removeScripType(activeScriptType);
+        this.extensionScript.removeScripType(passiveScriptType);
+    }
+
+    private ZapMenuItem getFrontEndScannerMenuToggle() {
+        if (menuFrontEndScanner == null) {
+            menuFrontEndScanner = new ZapMenuItem(PREFIX + ".topmenu.tools.title");
+
+            menuFrontEndScanner.addActionListener(new java.awt.event.ActionListener() {
+                @Override
+                public void actionPerformed(java.awt.event.ActionEvent ae) {
+                    frontEndScannerEnabled = !frontEndScannerEnabled;
+
+                    String value = frontEndScannerEnabled ? "on" : "off";
+                    View.getSingleton().showMessageDialog(
+                        Constant.messages
+                            .getString(ExtensionFrontEndScanner.PREFIX + ".topmenu.tools.msg", value));
+                }
+            });
+        }
+        return menuFrontEndScanner;
+    }
+
+    @Override
+    public String getAuthor() {
+        return Constant.ZAP_TEAM;
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(PREFIX + ".desc");
+    }
+
+    @Override
+    public URL getURL() {
+        try {
+            return new URL(Constant.ZAP_EXTENSIONS_PAGE);
+        } catch (MalformedURLException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public boolean onHttpRequestSend(HttpMessage msg) {
+        return true;
+    }
+
+    @Override
+    public boolean onHttpResponseReceive(HttpMessage msg) {
+        if (frontEndScannerEnabled && msg.getResponseHeader().isHtml()) {
+            try {
+                String html = msg.getResponseBody().toString();
+
+                Source document = new Source(html);
+                List<Element> heads = document.getAllElements("head");
+                Element head = heads.isEmpty() ? null : heads.get(0);
+
+                if (head != null && msg.getHistoryRef() != null) {
+                    String host = msg.getRequestHeader().getHeader("host");
+                    String frontEndApiUrl = API
+                        .getInstance()
+                        .getCallBackUrl(this.api, "https://" + host);
+
+                    int historyReferenceId = msg.getHistoryRef().getHistoryId();
+
+                    StringBuilder injectedContentBuilder = new StringBuilder()
+                        .append("<script type='text/javascript'>")
+                        .append("var frontEndScanner=(function() {")
+                        .append("const HISTORY_REFERENCE_ID = " + historyReferenceId + ";")

```

Comment:
It should append instead, .e.g.:
```Java
.append("const HISTORY_REFERENCE_ID = ").append(historyReferenceId).append(';')
```
(same for other inlined concatenations).

Reply:
Tried it and found it unreadable...